File: lib/src/main/java/hevs/graphics/TurtleGraphics.java
package hevs.graphics;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Point;
import java.awt.event.MouseMotionListener;

/**
 * Graphics class that emulates the tortoise in the Logo programming language
 * The turtle starts looking up with a black color, pen up
 * 
 * Basic port implementation by Pierre Roduit, rewritten for use with {@link FunGraphics}
 * by Pierre-André Mudry.
 * 
 * @see <a href="http://en.wikipedia.org/wiki/Turtle_graphics">Wikipedia
 *      description of Turtle Graphics</a>
 * @author <a href='mailto:pandre.mudry&#64;hevs.ch'> Pierre-Andre Mudry</a>
 * @version 2.01 Using {@link FunGraphics} as main class
 */
public class TurtleGraphics extends FunGraphics {

	private int x = fWidth / 2;
	private int y = fHeight / 2;
	private boolean penDown = false;
	private double angle = -Math.PI / 2.0; // Current rotation
	private Color color = Color.black;

	public TurtleGraphics(int width, int height, String windowName) {
		super(width, height, windowName, true);
		this.checkBorders = false;
	}

	public TurtleGraphics(int width, int height) {
		this(width, height, null);
	}

	/**
	 * Start the writing
	 */
	public void penDown() {
		penDown = true;
		// Write the pixel corresponding to the position
		Color oldColor = g2d.getColor();
		setColor(color);
		setPixel(x, y);
		setColor(oldColor);
	}

	/**
	 * Change the color of drawing
	 * 
	 * @param color
	 */
	public void changeColor(Color color) {
		this.color = color;
	}

	/**
	 * Stop the drawing
	 */
	public void penUp() {
		penDown = false;
	}

	/**
	 * Go forward the specified distance (writing if the pen is down)
	 * 
	 * @param distance
	 *            The distance to move
	 */
	public void forward(double distance) {
		// Compute new position
		int newX = x + (int) Math.round(Math.cos(angle) * distance);
		int newY = y + (int) Math.round(Math.sin(angle) * distance);

		// Write if the pen is down
		if (penDown) {
			Color oldColor = g2d.getColor();
			setColor(color);
			drawLine(x, y, newX, newY);
			setColor(oldColor);
		}
		x = newX;
		y = newY;
	}

	/**
	 * Jump to a specific location (without drawing)
	 * 
	 * @param x
	 *            X coordinate of the destination
	 * @param y
	 *            Y coordinate of the destination
	 */
	public void jump(int x, int y) {
		this.x = x;
		this.y = y;

		// If the pen is down, draw a point at destination
		if (penDown) {
			Color oldColor = g2d.getColor();
			setColor(color);
			setPixel(x, y);
			setColor(oldColor);
		}
	}

	/**
	 * @return The location of the turtle
	 */
	public Point getPosition() {
		return new Point(x, y);
	}

	/**
	 * Sets the width of the pen
	 * 
	 * @param w
	 */
	public void setWidth(float w) {
		g2d.setStroke(new BasicStroke(w));
	}

	/**
	 * @return The current turtle angle (in degrees)
	 */
	public double getTurtleAngle() {
		return (this.angle * 180.0 / Math.PI);
	}

	/**
	 * Turn the direction of writing with the specified angle
	 * 
	 * @param angle
	 *            specified angle in degrees
	 */
	public void turn(double angle) {
		this.angle += angle * Math.PI / 180;
	}

	/**
	 * Set the direction of writing to the specified angle
	 * 
	 * @param angle
	 *            specified angle in degrees
	 */
	public void setAngle(double angle) {
		this.angle = angle * Math.PI / 180;
	}

	/**
	 * Turn the direction of writing with the specified angle
	 * 
	 * @param angle
	 *            specified angle in radians
	 */
	public void turnRad(double angle) {
		this.angle += angle;
	}

	/**
	 * Set the direction of writing to the specified angle
	 * 
	 * @param angle
	 *            specified angle in radians
	 */
	public void setAngleRad(double angle) {
		this.angle = angle;
	}

	/** 
	 * @return The current turtle angle in radians
	 */
	public double getTurtleAngleRad() {
		return this.angle;
	}

	public void setMouseMotionManager(MouseMotionListener mouseMotionListener) {
		this.mainFrame.addMouseMotionListener(mouseMotionListener);
	}

}

File: src/main/scala/hevs/graphics/TurtleGraphics.scala
package hevs.graphics

import java.awt.{BasicStroke, Color, Point}
import java.awt.event.MouseMotionListener

/**
 * Graphics class that emulates the tortoise in the Logo programming language
 *
 * The turtle starts at the center of the window, looking up with a black color and pen down
 *
 * Basic port implementation by Pierre Roduit, rewritten for use with [[FunGraphics]]
 * by Pierre-André Mudry.
 *
 * @see <a href="http://en.wikipedia.org/wiki/Turtle_graphics">Wikipedia
 *      description of Turtle Graphics</a>
 * @author <a href='mailto:pandre.mudry&#64;hevs.ch'> Pierre-Andre Mudry</a>
 * @version 2.01 Using [[FunGraphics]] as main class
 *
 * @constructor Creates a turtle window
 * @param width        Width of the window
 * @param height       Height of the window
 * @param windowName   Title of the window
 */
class TurtleGraphics(width: Int, height: Int, windowName: String) extends FunGraphics(width, height, windowName, true) {
  this.checkBorders = false
  private var x = fWidth / 2.0
  private var y = fHeight / 2.0
  private var pDown = true
  private var angle = -Math.PI / 2.0 // Current rotation

  /**
   * Creates a turtle window
   *
   * @param width  Width of the window
   * @param height Height of the window
   */
  def this(width: Int, height: Int) = {
    this(width, height, null)
  }

  /**
   * Rounds a Double value and converts it to Int
   * @param a the value to round
   * @return the rounded value
   */
  private def round(a: Double) = a.round.toInt

  /**
   * Start the writing
   */
  def penDown(): Unit = {
    pDown = true
    // Write the pixel corresponding to the position
    setPixel(round(x), round(y))
  }

  /**
   * Change the color of drawing
   *
   * @param color the new color
   */
  def changeColor(color: Color): Unit = {
    setColor(color)
  }

  /**
   * Stop the drawing
   */
  def penUp(): Unit = {
    pDown = false
  }

  /**
   * Go forward the specified distance (writing if the pen is down)
   *
   * @param distance
   * The distance to move
   */
  def forward(distance: Double): Unit = { // Compute new position
    val newX = x + Math.cos(angle) * distance
    val newY = y + Math.sin(angle) * distance
    // Write if the pen is down
    if (pDown) drawLine(round(x), round(y), round(newX), round(newY))
    x = newX
    y = newY
  }

  /**
   * Jump to a specific location (without drawing)
   *
   * @param x
   * X coordinate of the destination
   * @param y
   * Y coordinate of the destination
   */
  def jump(x: Int, y: Int): Unit = {
    this.x = x
    this.y = y
    // If the pen is down, draw a point at destination
    if (pDown) setPixel(x, y)
  }

  /**
   * Gets the turtle's current position
   * @return The location of the turtle
   */
  def getPosition(): Point = new Point(round(x), round(y))

  /**
   * Sets the width of the pen
   *
   * @param w the new width
   */
  def setWidth(w: Float): Unit = {
    g2d.setStroke(new BasicStroke(w))
  }

  /**
   * Gets the turtle's current angle
   * @return The current turtle angle (in degrees)
   *         Angle 0 is east (right). A positive angle is clockwise.
   */
  def getTurtleAngle(): Double = this.angle * 180.0 / Math.PI

  /**
   * Turn the direction of writing with by specified angle
   * A positive angle is clockwise.
   *
   * @param angle specified angle in degrees
   */
  def turn(angle: Double): Unit = {
    this.angle += angle * Math.PI / 180
  }

  /**
   * Set the direction of writing to the specified angle.
   * Angle 0 is east (right). A positive angle is clockwise.
   *
   * @param angle
   * specified angle in degrees
   */
  def setAngle(angle: Double): Unit = {
    this.angle = angle * Math.PI / 180
  }

  /**
   * Turn the direction of writing by the specified angle.
   * A positive angle is clockwise.
   *
   * @param angle
   * specified angle in radians
   */
  def turnRad(angle: Double): Unit = {
    this.angle += angle
  }

  /**
   * Set the direction of writing to the specified angle.
   * Angle 0 is east (right). A positive angle is clockwise.
   *
   * @param angle
   * specified angle in radians
   */
  def setAngleRad(angle: Double): Unit = {
    this.angle = angle
  }

  /**
   * Get the turtle angle.
   * Angle 0 is east (right). A positive angle is clockwise.
   *
   * @return The current turtle angle in radians
   */
  def getTurtleAngleRad(): Double = this.angle

  /**
   * Adds a [[MouseMotionListener]] to the window to react on mouse movements
   * @param mouseMotionListener the [[MouseMotionListener]]
   */
  def setMouseMotionManager(mouseMotionListener: MouseMotionListener): Unit = {
    this.mainFrame.addMouseMotionListener(mouseMotionListener)
  }
}